rules/no-array-for-each.js

```diff
@@ -0,0 +1,284 @@
+			shouldAddParenthesesToExpressionStatementExpression(returnStatement.argument);
+		if (shouldAddParentheses) {
+			textBefore = '(';
+			textAfter = ')';
+		}
+
+		const shouldAddSemicolonBefore = needsSemicolon(previousToken, sourceCode, shouldAddParentheses ? '(' : nextToken.value);
+		if (shouldAddSemicolonBefore) {
+			textBefore = `;${textBefore}`;
+		}
+
+		if (textBefore) {
+			yield fixer.insertTextBefore(nextToken, textBefore);
+		}
+
+		if (textAfter) {
+			yield fixer.insertTextAfter(returnStatement.argument, textAfter);
+		}
+
+		// If `returnStatement` has no semi
+		const lastToken = sourceCode.getLastToken(returnStatement);
+		yield fixer.insertTextAfter(
+			returnStatement,
+			`${isSemicolonToken(lastToken) ? '' : ';'} continue;`
+		);
+	}
+
+	const shouldRemoveExpressionStatementLastToken = (token) => {
+		if (!isSemicolonToken(token)) {
+			return false;
+		}
+
+		if (callback.body.type === 'BlockStatement') {
+			return true;
+		}
+
+		const nextToken = sourceCode.getTokenAfter(token);
+		if (nextToken && needsSemicolon(token, sourceCode, nextToken.value)) {
+			return false;
+		}
+
+		return true;
+	};
+
+	return function * (fixer) {
+		yield fixer.replaceTextRange(getForOfLoopHeadRange(), getForOfLoopHeadText());
+
+		// Remove call expression trailing comma
+		const [penultimateToken, lastToken] = sourceCode.getLastTokens(callExpression, 2);
+		if (isCommaToken(penultimateToken)) {
+			yield fixer.remove(penultimateToken);
+		}
+
+		yield fixer.remove(lastToken);
+
+		for (const returnStatement of returnStatements) {
+			yield * replaceReturnStatement(returnStatement, fixer);
+		}
+
+		const expressionStatementLastToken = sourceCode.getLastToken(callExpression.parent);
+		if (shouldRemoveExpressionStatementLastToken(expressionStatementLastToken)) {
+			yield fixer.remove(expressionStatementLastToken, fixer);
+		}
+	};
+}
+
+function isFixable(callExpression, sourceCode, functionReturnStatements) {
+	// Check `CallExpression`
+	if (
+		callExpression.optional ||
+		isParenthesized(callExpression, sourceCode) ||
+		callExpression.arguments.length !== 1
+	) {
+		return false;
+	}
+
+	// Check `CallExpression.parent`
+	if (callExpression.parent.type !== 'ExpressionStatement') {
+		return false;
+	}
+
+	// Check `CallExpression.callee`
+	if (callExpression.callee.optional) {
+		return false;
+	}
+
+	// Check `CallExpression.arguments[0]`;
+	const [callback] = callExpression.arguments;
+	if (
+		// Leave non-function type to `no-array-callback-reference` rule
+		(callback.type !== 'FunctionExpression' && callback.type !== 'ArrowFunctionExpression') ||
+    callback.async ||
+    callback.generator
+	) {
+		return false;
+	}
+
+	// Check `callback.params`
+	const parameters = callback.params;
+	if (
+		!(parameters.length === 1 || parameters.length === 2) ||
+		parameters.some(parameter => parameter.type !== 'Identifier')
+	) {
+		return false;
+	}
+
+	// TODO: check parameters conflicts
```

Comment:
Not really sure how to check this,

Should not fix



```js
foo.forEach(foo => bar());
````

->

```js
for (const foo of foo) bar();
````



```js
a[foo].forEach(foo => bar());
````

->

```js
for (const foo of a[foo]) bar();
````

-----

Should fix




```js
a((foo) => foo).forEach(foo => bar());
````
->
```js
for (const foo of a((foo) => foo)) bar();
````

Reply:
Figured a way to detect this, but not very efficient.

Reply:
A similar logic maybe possible to auto fix destructuring parameters, but I'm going to try in a seperate PR.